velero-plugin-for-microsoft-azure/object_store.go

```diff
@@ -51,190 +48,207 @@ type containerGetter interface {
-	return b.blob.Get(options)
+func (b *azureBlob) Get(options *azblob.BlobDownloadOptions) (azblob.BlobDownloadResponse, error) {
+	return b.blobClient.Download(context.TODO(), options)
 }
 
-func (b *azureBlob) Delete(options *storage.DeleteBlobOptions) error {
-	return b.blob.Delete(options)
+func (b *azureBlob) Delete(options *azblob.BlobDeleteOptions) error {
+	_, err := b.blobClient.Delete(context.TODO(), options)
+	return err
 }
 
-func (b *azureBlob) GetSASURI(options *storage.BlobSASOptions) (string, error) {
-	return b.blob.GetSASURI(*options)
+// When the sharedKeyCredential is provided service SAS is used else delegation SAS is used
+func (b *azureBlob) GetSASURI(storageEndpointSuffix, storageAccount string, ttl time.Duration, sharedKeyCredential *azblob.SharedKeyCredential) (string, error) {
+	// because of clock skew it can happen that the token is not yet valid, so make it valid in the past
+	start := time.Now().Add(-10 * time.Minute)
+	expiry := time.Now().Add(ttl)
+	if sharedKeyCredential == nil {
+		userDelegationKey, err := GetUserDelegationKey(storageEndpointSuffix, storageAccount, start, expiry)
+		if err != nil {
+			return "", err
+		}
+
+		url, err := CreateSignedURL(storageEndpointSuffix, storageAccount, b.container, b.blob, userDelegationKey, start, expiry)
+		if err != nil {
+			return "", err
+		}
+		return url, nil
+	}
+
+	queryParam, err := azblob.BlobSASSignatureValues{
+		ContainerName: b.container,
+		BlobName:      b.blob,
+		Permissions:   azblob.BlobSASPermissions{Read: true}.String(),
+		ExpiryTime:    time.Now().UTC().Add(ttl),
+	}.NewSASQueryParameters(sharedKeyCredential)
+
+	if err != nil {
+		return "", err
+	}
+	url := fmt.Sprintf("%s?%s", b.blobClient.URL(), queryParam.Encode())
+	return url, nil
 }
 
 type ObjectStore struct {
-	log             logrus.FieldLogger
-	containerGetter containerGetter
-	blobGetter      blobGetter
-	blockSize       int
+	log logrus.FieldLogger
+
+	containerGetter       containerGetter
+	blobGetter            blobGetter
+	blockSize             int
+	storageAccount        string
+	storageEndpointSuffix string
+	// we need to keep the credential here to create the sas url
+	sharedKeyCredential *azblob.SharedKeyCredential
 }
 
 func newObjectStore(logger logrus.FieldLogger) *ObjectStore {
 	return &ObjectStore{log: logger}
 }
 
-// getSubscriptionID gets the subscription ID from the 'config' map if it contains
-// it, else from the AZURE_SUBSCRIPTION_ID environment variable.
-func getSubscriptionID(config map[string]string) string {
-	if subscriptionID := config[subscriptionIDConfigKey]; subscriptionID != "" {
-		return subscriptionID
-	}
-
-	return os.Getenv(subscriptionIDEnvVar)
-}
-
-func getStorageAccountKey(config map[string]string) (string, *azure.Environment, error) {
+// get storage account key from env var whose name is in config[storageAccountKeyEnvVarConfigKey].
+func getStorageKey(config map[string]string) (string, error) {
```

Comment:
How about renaming the function name to `getStorageAccountAccessKey`? It's confused for me what `StorageKey` is?

Reply:
I renamed it to getStorageAccountKey to match the variable names